core/common/src/main/java/alluxio/util/ConfigurationUtils.java

```diff
@@ -247,4 +271,160 @@ public static boolean masterHostConfigured() {
+   */
+  public static synchronized void reloadProperties() {
+    // Step1: bootstrap the configuration. This is necessary because we need to resolve alluxio.home
+    // (likely to be in system properties) to locate the conf dir to search for the site property
+    // file.
+    AlluxioProperties properties = new AlluxioProperties();
+    InstancedConfiguration conf = new InstancedConfiguration(properties);
+    properties.merge(System.getProperties(), Source.SYSTEM_PROPERTY);
+    Properties siteProps = null;
+
+    // Step2: Load site specific properties file if not in test mode. Note that we decide whether in
+    // test mode by default properties and system properties (via getBoolean).
+    if (conf.getBoolean(PropertyKey.TEST_MODE)) {
+      conf.validate();
+      sDefaultProperties = properties;
+      return;
+    }
+
+    // we are not in test mode, load site properties
+    String confPaths = conf.get(PropertyKey.SITE_CONF_DIR);
+    String[] confPathList = confPaths.split(",");
+    String sitePropertyFile = ConfigurationUtils
+            .searchPropertiesFile(Constants.SITE_PROPERTIES, confPathList);
+    if (sitePropertyFile != null) {
+      siteProps = loadPropertiesFromFile(sitePropertyFile);
+      sSourcePropertyFile = sitePropertyFile;
+    } else {
+      URL resource =
+          ConfigurationUtils.class.getClassLoader().getResource(Constants.SITE_PROPERTIES);
+      if (resource != null) {
+        siteProps = loadPropertiesFromResource(resource);
+        if (siteProps != null) {
+          sSourcePropertyFile = resource.getPath();
+        }
+      }
+    }
+    properties.merge(siteProps, Source.siteProperty(sSourcePropertyFile));
+    conf.validate();
+    sDefaultProperties = properties;
+  }
+
+  /**
+   * Merges the current configuration properties with new properties. If a property exists
+   * both in the new and current configuration, the one from the new configuration wins if
+   * its priority is higher or equal than the existing one.
+   *
+   * @param conf the base configuration
+   * @param properties the source {@link Properties} to be merged
+   * @param source the source of the the properties (e.g., system property, default and etc)
+   */
+  public static void  merge(AlluxioConfiguration conf, Map<?, ?> properties,
+      Source source) {
+    AlluxioProperties props = conf.getProperties();
+    props.merge(properties, source);
+  }
+
+  /**
+   * Loads cluster default values from the meta master.
+   *
+   * @param address the master address
+   * @param conf configuration to use
+   * @return An updated configuration object containing the cluster defaults
+   */
+  public static AlluxioConfiguration loadClusterDefaults(InetSocketAddress address,
+      AlluxioConfiguration conf)
+      throws AlluxioStatusException {
+    if (!conf.getBoolean(PropertyKey.USER_CONF_CLUSTER_DEFAULT_ENABLED)
+        || conf.clusterDefaultsLoaded()) {
+      return conf;
+    }
+    if (conf.clusterDefaultsLoaded()) {
```

Comment:
this is already checked earlier? Is this supposed to be thread-safe? It looks like 2 different threads will be able to load cluster defaults on the same configuration object.

Reply:
I will remove the second check however